### ISSUE TITLE
RFC: Replacement for Buffer engine in MergeTree

### ISSUE THREAD

**Issue by azat** [ClickHouse/ClickHouse#40728]
Right now there are 3 types of parts in `MergeTree`:

- `Wide`
- `Compact`
- `InMemory`

Each next should be faster the the previous.

I did some benchmarks and here what I got:

*But note, that this benchmark is bad, real numbers will be worse, especially for `ReplicatedMergeTree`.*

*And I also have other numbers (and some experiments on how to improve the speed of InMemory parts but this break some assumptions in the code), but I don't want to pollute the description with them*

table type|qps
-|-
mt_wide                        | 684.201
mt_compact                     | 792.943
**mt_inmemory_no_wal**             | **916.427**
mt_inmemory_wal                | 902.082
mt_inmemory_wal_fsync          | 881.463
**rep_mt_inmemory_no_wal**         | **362.900**
rep_mt_inmemory_no_wal_sync_zk | 355.636
rep_mt_inmemory_wal            | 367.033
rep_mt_inmemory_wal_fsync      | 338.189
**buffer**                         | **1631.094**

<details>

<summary>Where</summary>

- `rep_` - `ReplicatedMergeTree`
- `mt` - `MergeTree`
- `fsync` - `write_ahead_log_bytes_to_fsync=1`
- `wal` - `in_memory_parts_enable_wal=1`

</details>

So as you can see `InMemory` parts for `ReplicatedMergeTree` is the most slowest, and this is obvious, because it requires ZooKeeper. And it will also increase the load to the ZooKeeper.

**That said that obviously you cannot use `InMemory` as a replacement for `Buffer` engine right now.**

So what I'm suggesting is one of the following:

- Either add an option to avoid committing `InMemory` parts to ZooKeeper (and kind of make sense, especially if you don't have WAL enabled, yes I know that they will be flushed on shutdown, but what if server died?). **But I'm afraid that this can break some guarantees in the code and will require lots of special cases.**

- Or maybe some `Temporary` (thought this conflicts with the part state, but this is just a working title) parts:
  - `SELECT` will not handle them
  - Will not be committed to ZooKeeper
  - Will not have checksums - in my testing it has ~33% overhead
  - Will not calculate any indexes
  - Will not have disk reservations - right `InMemory` has them since they can be committed to disk at shutdown, thought this can be significantly postponed
  - Will have separate WAL for restoring
  - But again this can have some special cases in the `ReplicatedMergeTree` code...

- Name your...

But instead of implementing one of the following, I decided to post an RFC first.

**So if someone has some thoughts about this (maybe @tavplubix for `ReplicatedMergeTree` and only local parts there), please go ahead!**

**Comment by alexey-milovidov** [comment]
Sounds like you need async INSERTs with WAL.

**Comment by azat** [comment]
>Sounds like you need async INSERTs with WAL.

And it also stores the unparsed query.

Also I though about having something on per-partition basis, but maybe this will not have any benefits.

So maybe you right (plus I do dislike the idea of non-replicated parts in replicated engine).

**Comment by azat** [comment]
The problem with async INSERT's is that it does not make sense for INSERT SELECT (at least right now).

Plus I've actually tried to overcome this with a huge Buffer table, but this is not the same as Buffer-per-part, works worse, especially in case of ReplicatedMergeTree.

So I've tried to go with separate part type for this - `Buffer`. But I still dislike that they are not commited to ZooKeeper.

So here is a rfc-draft patch for this crazy idea - #44056
